File: code/Algorithm/leetcode/2019.11/234.回文链表.cpp
/*
 * @lc app=leetcode.cn id=234 lang=cpp
 *
 * [234] 回文链表
 *
 * https://leetcode-cn.com/problems/palindrome-linked-list/description/
 *
 * algorithms
 * Easy (39.01%)
 * Likes:    315
 * Dislikes: 0
 * Total Accepted:    48.5K
 * Total Submissions: 123.7K
 * Testcase Example:  '[1,2]'
 *
 * 请判断一个链表是否为回文链表。
 * 
 * 示例 1:
 * 
 * 输入: 1->2
 * 输出: false
 * 
 * 示例 2:
 * 
 * 输入: 1->2->2->1
 * 输出: true
 * 
 * 
 * 进阶：
 * 你能否用 O(n) 时间复杂度和 O(1) 空间复杂度解决此题？
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode(int x) : val(x), next(nullptr) {}
// };
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        if (head == nullptr || head->next == nullptr)
            return true;
        
        // 通过快慢节点找到中点
        ListNode *fast = head;
        ListNode *slow = head;
        while (fast->next != nullptr)
        {
            if (fast->next->next == nullptr)
            {
                fast = fast->next;
                break;
            }
            else
            {
                fast = fast->next->next;
            }
            slow = slow->next;
        }
        // 将后半部分链表倒序
        ListNode *p = slow->next;
        ListNode *nex;
        ListNode *pre = nullptr;
        while (p->next != nullptr)
        {
            nex = p->next;

            p->next = pre;
            pre = p;

            p = nex;
        }
        p->next = pre;
        // 开始判断回文
        ListNode *left = head;
        ListNode *right = p;
        while (left != nullptr && right != nullptr)
        {
            if (left->val != right->val)
                return false;
            left = left->next;
            right = right->next;
        }
        return true;
    }
};
// @lc code=end

File: code/JavaLab/src/main/java/algorithm/leetcode/linklist_25/234.回文链表.java
package algorithm.leetcode.linklist_25;

import algorithm.config.ListNode;

/**
 * https://leetcode-cn.com/problems/palindrome-linked-list/
 * <p>
 * 给你一个单链表的头节点 head ，请你判断该链表是否为回文链表。如果是，返回 true ；否则，返回 false 。
 */
class LC234 {


    public boolean isPalindrome(ListNode head) {

        // 特判一下
        if (head == null || head.next == null) {
            return true;
        }
        if (head.next.next == null) {
            return head.val == head.next.val;
        }


        // 快慢指针找到中点p1
        ListNode p1 = head, p2 = head;
        ListNode pre = p1;
        while (p2 != null && p2.next != null) {
            pre = p1;
            p1 = p1.next;
            p2 = p2.next.next;
        }
        // 从中间断开链表
        pre.next = null;
        // 翻转链表
        ListNode head2 = reverse(p1);

        while (head != null && head2 != null) {
            if (head.val != head2.val) {
                return false;
            }
            head = head.next;
            head2 = head2.next;
        }
        return true;
    }

    public ListNode reverse(ListNode head) {

        if (head == null || head.next == null) {
            return null;
        }

        ListNode pre = null;
        ListNode p = head;
        while (p != null) {
            ListNode next = p.next;
            p.next = pre;
            pre = p;
            p = next;
        }
        return pre;
    }
}